File: web/js/providers/mn_state.js
// Port of backend/services/mn_state.py
import { BaseParcelProvider } from "./base.js";

// All seven metro counties are published on one MN state ArcGIS server, one
// layer each (0 Anoka, 1 Carver, 2 Dakota, 3 Hennepin, 4 Ramsey, 5 Scott,
// 6 Washington). Anoka/Hennepin/Ramsey/Scott use their own county services;
// the ones below share this statewide schema.
export const MN_STATE_PARCELS_FEATURESERVER =
  "https://arcgis.metc.state.mn.us/data1/rest/services/parcels/Parcels/FeatureServer";

export class MnStateSchemaProvider extends BaseParcelProvider {
  // Shared adapter for metro counties on the MN state parcel server.
  // Schema: PIN / OWNER_NAME / ANUMBER (int) / ST_NAME / ST_POS_TYP /
  // ST_POS_DIR / CTU_NAME / ZIP. Subclasses pin the layer index, source
  // label, and a CO_NAME safety clause so a mis-routed query can never
  // return another county's parcels.
  coName = "";
  parcelIdField = "PIN";
  ownerField = "OWNER_NAME";
  addressField = "";
  extraOutFields = [
    "ANUMBER",
    "ST_NAME",
    "ST_PRE_TYP",
    "ST_PRE_DIR",
    "ST_POS_TYP",
    "ST_POS_DIR",
    "CTU_NAME",
    "ZIP",
  ];
  adjacentSpatialRel = "esriSpatialRelIntersects";

  async _queryCounty(params, budget) {
    if (this.coName) {
      params = { ...params };
      const where = params.where || "1=1";
      params.where = `(${where}) AND CO_NAME = '${this._sqlEscape(this.coName)}'`;
    }
    return super._queryCounty(params, budget);
  }

  _getAddressWhereExact(cleaned) {
    const houseNo = this._extractHouseNumber(cleaned);
    const streetToken = this._extractStreetToken(cleaned);
    if (streetToken === null && houseNo === null) return "1=0";
    const clauses = [];
    if (houseNo !== null) clauses.push(`ANUMBER = ${houseNo}`);
    clauses.push(`UPPER(ST_NAME) LIKE '%${this._sqlEscape(streetToken)}%'`);
    return clauses.join(" AND ");
  }

  _getAddressWhereContains(cleaned) {
    const streetToken = this._extractStreetToken(cleaned);
    if (!streetToken) return "1=0";
    return `UPPER(ST_NAME) LIKE '%${this._sqlEscape(streetToken)}%'`;
  }

  _buildAddress(attrs) {
    const number = this._formatInt(attrs.ANUMBER);
    const streetName = String(attrs.ST_NAME ?? "").trim();
    const posType = String(attrs.ST_POS_TYP ?? "").trim();
    const posDir = String(attrs.ST_POS_DIR ?? "").trim();
    const city = String(attrs.CTU_NAME ?? "").trim();
    const zip = String(attrs.ZIP ?? "").trim();

    const parts = [];
    if (number) parts.push(number);
    if (streetName) parts.push(streetName);
    if (posType) parts.push(posType);
    if (posDir) parts.push(posDir);
    const streetLine = parts.join(" ");

    const suffixParts = [];
    suffixParts.push(city || "MN");
    if (zip) suffixParts.push(zip);
    const suffix = suffixParts.join(" ");

    if (streetLine && suffix) return `${streetLine}, ${suffix}`;
    return streetLine || suffix;
  }

  _formatInt(value) {
    if (value === null || value === undefined) return "";
    const n = parseInt(value, 10);
    if (!Number.isFinite(n)) return String(value).trim();
    return n > 0 ? String(n) : "";
  }

  _extractHouseNumber(address) {
    const firstPart = address.split(",")[0].trim();
    const m = firstPart.match(/^(\d+)/);
    return m ? parseInt(m[1], 10) : null;
  }

  _extractStreetToken(address) {
    const withoutHouse = address.replace(/^\s*\d+\S?\s+/, "").trim();
    if (!withoutHouse) return null;
    const first = withoutHouse.split(/\s+/)[0].trim();
    return first || null;
  }
}

export class DakotaProvider extends MnStateSchemaProvider {
  endpointUrl = MN_STATE_PARCELS_FEATURESERVER + "/2/query";
  sourceLabel = "dakota_county_arcgis_state";
  coName = "Dakota";
}

export class WashingtonProvider extends MnStateSchemaProvider {
  endpointUrl = MN_STATE_PARCELS_FEATURESERVER + "/6/query";
  sourceLabel = "washington_county_arcgis_state";
  coName = "Washington";
}

export class CarverProvider extends MnStateSchemaProvider {
  // Verified 2026-08-21: OWNER_NAME is 100% null in the Carver layer, while
  // TAX_NAME (the taxpayer name) is populated for 47,585 of 47,886 parcels.
  ownerField = "TAX_NAME";
  endpointUrl = MN_STATE_PARCELS_FEATURESERVER + "/1/query";
  sourceLabel = "carver_county_arcgis_state";
  coName = "Carver";
}

File: backend/services/mn_state.py
from __future__ import annotations

import re
from typing import Any

from backend.services.base import BaseParcelService, RequestBudget


# All seven metro counties are published on one MN state ArcGIS server, one
# layer each (0 Anoka, 1 Carver, 2 Dakota, 3 Hennepin, 4 Ramsey, 5 Scott,
# 6 Washington). Anoka/Hennepin/Ramsey/Scott use their own county services;
# the ones below share this statewide schema.
MN_STATE_PARCELS_FEATURESERVER = (
    "https://arcgis.metc.state.mn.us/data1/rest/services/parcels/Parcels/FeatureServer"
)


class MnStateSchemaParcelService(BaseParcelService):
    """Shared adapter for metro counties on the MN state parcel server.

    Schema: PIN / OWNER_NAME / ANUMBER (int) / ST_NAME / ST_POS_TYP /
    ST_POS_DIR / CTU_NAME / ZIP. Subclasses pin the layer index, source
    label, and a CO_NAME safety clause so a mis-routed query can never
    return another county's parcels.
    """

    co_name: str = ""
    parcel_id_field = "PIN"
    owner_field = "OWNER_NAME"
    address_field = ""
    extra_out_fields = [
        "ANUMBER",
        "ST_NAME",
        "ST_PRE_TYP",
        "ST_PRE_DIR",
        "ST_POS_TYP",
        "ST_POS_DIR",
        "CTU_NAME",
        "ZIP",
    ]
    adjacent_spatial_rel = "esriSpatialRelIntersects"

    async def _query_county(
        self,
        params: dict[str, str],
        *,
        budget: RequestBudget,
    ) -> list[dict[str, Any]]:
        if self.co_name:
            params = dict(params)
            where = params.get("where") or "1=1"
            params["where"] = (
                f"({where}) AND CO_NAME = '{self._sql_escape(self.co_name)}'"
            )
        return await super()._query_county(params, budget=budget)

    def _get_address_where_exact(self, cleaned: str) -> str:
        house_no = self._extract_house_number(cleaned)
        street_token = self._extract_street_token(cleaned)
        if street_token is None and house_no is None:
            return "1=0"
        clauses = []
        if house_no is not None:
            clauses.append(f"ANUMBER = {house_no}")
        clauses.append(f"UPPER(ST_NAME) LIKE '%{self._sql_escape(street_token)}%'")
        return " AND ".join(clauses)

    def _get_address_where_contains(self, cleaned: str) -> str:
        street_token = self._extract_street_token(cleaned)
        if not street_token:
            return "1=0"
        return f"UPPER(ST_NAME) LIKE '%{self._sql_escape(street_token)}%'"

    def _build_address(self, attrs: dict[str, Any]) -> str:
        number = self._format_int(attrs.get("ANUMBER"))
        street_name = str(attrs.get("ST_NAME") or "").strip()
        pos_type = str(attrs.get("ST_POS_TYP") or "").strip()
        pos_dir = str(attrs.get("ST_POS_DIR") or "").strip()
        city = str(attrs.get("CTU_NAME") or "").strip()
        zip_str = str(attrs.get("ZIP") or "").strip()

        parts: list[str] = []
        if number:
            parts.append(number)
        if street_name:
            parts.append(street_name)
        if pos_type:
            parts.append(pos_type)
        if pos_dir:
            parts.append(pos_dir)
        street_line = " ".join(parts)

        suffix_parts: list[str] = []
        if city:
            suffix_parts.append(city)
        else:
            suffix_parts.append("MN")
        if zip_str:
            suffix_parts.append(zip_str)
        suffix = " ".join(suffix_parts)

        if street_line and suffix:
            return f"{street_line}, {suffix}"
        return street_line or suffix

    @staticmethod
    def _format_int(value: Any) -> str:
        if value is None:
            return ""
        try:
            as_int = int(value)
        except (TypeError, ValueError):
            return str(value).strip()
        return str(as_int) if as_int > 0 else ""

    @staticmethod
    def _extract_house_number(address: str) -> int | None:
        first_part = address.split(",", maxsplit=1)[0].strip()
        match = re.match(r"^(\d+)", first_part)
        if match:
            return int(match.group(1))
        return None

    @staticmethod
    def _extract_street_token(address: str) -> str | None:
        without_house = re.sub(r"^\s*\d+\S?\s+", "", address).strip()
        if not without_house:
            return None
        first = without_house.split()[0].strip()
        return first or None


class DakotaParcelService(MnStateSchemaParcelService):

    endpoint_url = MN_STATE_PARCELS_FEATURESERVER + "/2/query"
    source_label = "dakota_county_arcgis_state"
    co_name = "Dakota"


class WashingtonParcelService(MnStateSchemaParcelService):

    endpoint_url = MN_STATE_PARCELS_FEATURESERVER + "/6/query"
    source_label = "washington_county_arcgis_state"
    co_name = "Washington"


class CarverParcelService(MnStateSchemaParcelService):
    # Verified 2026-08-21: OWNER_NAME is 100% null in the Carver layer, while
    # TAX_NAME (the taxpayer name) is populated for 47,585 of 47,886 parcels.
    owner_field = "TAX_NAME"

    endpoint_url = MN_STATE_PARCELS_FEATURESERVER + "/1/query"
    source_label = "carver_county_arcgis_state"
    co_name = "Carver"
